### ISSUE TITLE
Release tarball with content of submodules

### ISSUE THREAD

**Issue by ghisvail** [arrayfire/arrayfire#712]
Dear ArrayFire developers,

It would simplify downstream packaging of ArrayFire a great deal if you guys could provide additional release tarballs with the content of the submodules. The tarballs generated from the git tags omit these, which means neither the documentation nor the test suite can be built correctly from source.

These additional release tarballs could be provided via Git releases maybe ?

Many thanks,
Ghis

**Comment by pavanky** [comment]
Hi Ghisvall,

We are infact trying to make a release this week (possibly today!). Please watch out.

**Comment by pavanky** [comment]
Also the submodules are only necessary for ~~building~~ **running** the examples and tests. Is this necessary for the packages ?

**Comment by ghisvail** [comment]
> We are infact trying to make a release this week (possibly today!).

Awesome ! Congrats !

> Submodules are only necessary for building running the examples and tests.  Is this necessary for the packages ?

Examples no. Test suite most certainly yes.

**Comment by shehzan10** [comment]
Correction: Examples do requires the assets submodule. These contain data like images, mnist data.

**Comment by pavanky** [comment]
@ghisvail I am new to the difficulties of packaging, so please bear with me for a bit.

I just realized another issue that might come into play. ArrayFire ~~depends~~ *might depend* on development version of clFFT and clBLAS for some bug fixes. This is alright if you are statically linking those libraries with ArrayFire.

Considering that @kif (https://github.com/kif) is working on integrating clFFT (and perhaps clBLAS) into debian, I assume you want to use the libraries from the repos as opposed to statically linking them.

This might lead to some bugs in ArrayFire that are only present on debian (or any other distro for that matter)  but not in our development branch.

So I guess my question is, what is the ideal situation to minimize situations like this. Can the upstream developers such as ourselves do something to reduce the burden on you ?

P.S. We also depend on many other libraries, but I am less worried about breakages coming from those libraries.

**Comment by ghisvail** [comment]
> Considering that @kif (https://github.com/kif) is working on integrating clFFT (and perhaps clBLAS) into debian

I actually took over Jerome for clFFT and I am also taking care of clBLAS ^^.

> I assume you want to use the libraries from the repos as opposed to statically linking them.

That would be the aim, having the packaged version of the ArrayFire OpenCL backend depend on the packaged version of clBLAS and clFFT. Dyanmic linking allows us to have a single place to fix bugs (the package itself).

> This might lead to some bugs in ArrayFire that are only present on debian repos (or any other distro for that matter) but not in our development branch.

Not quite sure what you mean here. Do you have an example in mind ?

> So I guess my question is, what is the ideal solution to minimize situations like this. Can the upstream developers such as ourselves do something to reduce the burden on you ?

Again not sure which situations you are referring to. Please be more explicit.

> P.S. We also depend on many other libraries, but I am less worried about breakages coming from those libraries. 

You mean API breakage ? Is that still a risk with clBLAS / clFFT ?

**Comment by pavanky** [comment]
> > Considering that @kif (https://github.com/kif) is working on integrating clFFT (and perhaps clBLAS) into debian

> I actually took over Jerome for clFFT and I am also taking care of clBLAS ^^.

Ah good to know. I just saw him interacting with the clFFT devs some time ago. Thanks for clearing it up.

> You mean API breakage ? Is that still a risk with clBLAS / clFFT ?

No the API is stable. The problems will be with accuracy and functionality on various OpenCL based devices.

Given the wide range of quirks and inconsistencies in OpenCL runtimes, clMathLibraries and arrayfire are broken on a few platforms. Sometimes it is the problem with the OpenCL runtimes, sometimes it is an issue with the libraries themselves. 

For example, we find a bug when using clFFT on intel CPUs, we fix it and it gets updated in their development branch. 

Ideally, when we are making a minor release of ArrayFire, we want this to be fixed. If clFFT does not do a release before we do, the version on debian still points to last stable release which has this bug. 

Perhaps this is a smaller issue than I anticipated. This can be fixed by cherry picking the appropriate commits to the mirror you are maintaining.

**Comment by ghisvail** [comment]
> Perhaps this is a smaller issue than I anticipated. This can be fixed by cherry picking the appropriate commits to the mirror you are maintaining.

Absolutely, we can (and will) cherry-pick commits to provide bug-fixed packages.

**Comment by pavanky** [comment]
About the original subject of this issue, is it OK to host the tarball on our CI server ? How often will it be hit with downloads?

**Comment by ghisvail** [comment]
The Debian BTS monitors new releases using a watch file (which defines a perl regexp to extract the upstream version from the URL where release tarballs are hosted). The BTS does *not* perform any download.

This tarball would be typically downloaded once per new release by the package maintainer and then be committed to our mirror in compressed form. As long as you guys provide a reliable location to fetch all tarballs...

**Comment by pavanky** [comment]
This should be possible. I think we will just put it on our main website as it has a more reliable server than our CI.

**Comment by ghisvail** [comment]
Can't you guys just do it via Github releases ?

**Comment by pavanky** [comment]
That would be ideal, but from what I can see I don't think we can include the submodules. If we can, we will just go down that route.

**Comment by pavanky** [comment]
Nevermind, I did not look hard enough. I think it is possible.

**Comment by ghisvail** [comment]
Maybe a stupid question, but why provide gtest as a submodule when you can do it with ExternalProject_Add as it is done in clBLAS / clFFT ?

ExternalProject_Add sounds to me like a more reasonable approach + it is easier to patch for using a system provided one like in Debian.

FYI, Debian and I believe does not actually provide a compiled libgtest (so find_gtest would be useless actually), since it is not recommended by Google, but does provide the gtest sources in a specific directory (/usr/src/gtest). Then, ExternalProject_Add can use them via
```
DOWNLOAD_COMMAND=""
SOURCE_DIR=/usr/src/test
```
instead of downloading it from upstream via the URL options, as done in clBLAS, clFFT.

**Comment by shehzan10** [comment]
Personally, I think it is fair to have gtest downloaded with the external build. I have wondered the same but never got around doing something about it.

With regards to using the system directory to build gtest, I think it is one too many options for the user, and probably redundant. The redundancy coming from the fact that any developer could build gtest in the system directory and simply use the USE_SYSTEM_GTEST option. In this case findgtest would work just fine. 

My question to you would be whether packagers (and other downstream projects) like yourself would use gtest downloaded by the external project or would you want it to build from /usr/src?

If you use gtest primarily downloaded by external project, then having it build from /usr/src would just be an additional option that most devs won't use. But the option would still be available using the USE_SYSTEM_GTEST option.

Let's see what Pavan has to say about this.

**Comment by ghisvail** [comment]
> The redundancy coming from the fact that any developer could build gtest in the system directory and simply use the USE_SYSTEM_GTEST option. In this case findgtest would work just fine.

Actually Google explictly discourage to do so:
https://code.google.com/p/googletest/wiki/V1_7_FAQ#Why_is_it_not_recommended_to_install_a_pre-compiled_copy_of_Goog

Which means USE_SYSTEM_GTEST / findgtest is sort of useless.

> My question to you would be whether packagers (and other downstream projects) like yourself would use gtest downloaded by the external project or would you want it to build from /usr/src?

We cannot use any remote resources at build time. However, this is not a problem with gtest because ExternalProject can use local source directories, using the two options I listed previously. So we would need to point ExternalProject to the gtest source instead of the remote URL, which is a very minimal patch (i.e. commenting the URL options and adding the SOURCE_DIR and DOWNLOAD_COMMAND options)

You guys actually don't need to do anything, besides providing a standard ExternalProject for gtest like the one in clBLAS really.

> If you use gtest primarily downloaded by external project, then having it build from /usr/src would just be an additional option that most devs won't use. But the option would still be available using the USE_SYSTEM_GTEST option.

I would just not care at that point, since it is very much limited to downstream usage. Just remove the USE_SYSTEM_GTEST option (because even Google discourages this practice) and provide a single ExternalProject_Add for gtest similar to clBLAS.

**Comment by ghisvail** [comment]
FYI, I validated this approach on clBLAS successfully.

**Comment by shehzan10** [comment]
As I said, I think its fine to have the build_gtest like clBLAS does. I am aware of google tests discouraging gtest from system. That's why we build it at compile time.

> We cannot use any remote resources at build time. However, this is not a problem with gtest because ExternalProject can use local source directories, using the two options I listed previously. So we would need to point ExternalProject to the gtest source instead of the remote URL, which is a very minimal patch (i.e. commenting the URL options and adding the SOURCE_DIR and DOWNLOAD_COMMAND options)

What I'm not clear about is do you want the source for that build to be downloaded from git or be from the source package (/usr/src) or both (depending on an option the user can set)?

**Comment by umar456** [comment]
@ghisvail with regard to gtest. See issue #424 
Gtest is being built as an external project. We are using the submodule directory as the source. We can update the [external project command](https://github.com/arrayfire/arrayfire/blob/devel/CMakeModules/build_gtest.cmake#L49) to use the debian source directory instead of the provided submodule if necessary.

**Comment by ghisvail** [comment]
@shehzan10 
> What I'm not clear about is do you want the source for that build to be downloaded from git or be from the source package (/usr/src)

We would use the content of /usr/src/gtest regardless of whether the content of test/gtest is provided in the release tarball.

@shehzan10 , @umar456 
> or both (depending on an option the user can set)?

Since this is Debian specific (other distros may handle gtest differently), I guess no further options should be provided. IMO, it is not worth for you guys to maintain yet another build option just for this. Even usage of USE_SYSTEM_GTEST seems pointless to me, based on what Google recommends.

With regards to #424, I am not quite sure why you guys went for submodules rather than what clBLAS does, but I am sure you have your reasons.

My apologize for the noise on the gtest stuff, I think I mistook the ArrayFire situation for the clFFT one whereby the test suite only relies on a system-wide gtest installation. My lesson for working on the packaging of all 3 at once this weekend ^^.

**Comment by pavanky** [comment]
We had gtest as an external module before. The problem was the repo we were trying to download from was highly unreliable and it would fail to download often. So we did two things simultaneously to address this problem.

1. Change the repo to a more reliable one
2. Move from external module to a submodule.

Now submodules have the benefit of being downloaded only once where as external modules are downloaded every time you do a clean build. This is the reason we ended up keeping it this way. Technically we could have just used (1), but it is nice to not download the repo over and over again.

**Comment by ghisvail** [comment]
FYI, regarding the issue, I can also provide a custom target in the downstream packaging which would be responsible for cloning the repository and its submodules and pack it to a release tarball, which would then be used for packaging. Let me know if that suits you guys better.

**Comment by pavanky** [comment]
@ghisvail when #750 is resolved, we won't have to worry about sub module dependencies. Hopefully this will be fixed today.

**Comment by ghisvail** [comment]
Brilliant

**Comment by pavanky** [comment]
@ghisvail Looks like we can't remove the submodule dependency n time for 3.0. We will manually upload the 3.0 with submodules. We will try to fix the issue before the next minor release.

**Comment by ghisvail** [comment]
> We will manually upload the 3.0 with submodules.

Sounds good to me. 

> We will try to fix the issue before the next minor release.

No pressure on my end :). You guys should take your time to find what works best for the project.

**Comment by pavanky** [comment]
@ghisvail github is limiting the archive to 10MB. We uploaded the source code with submodules to our website. You can find the link at the top of this page:

https://github.com/arrayfire/arrayfire/releases/tag/v3.0.0
